### PR TITLE
[enhancement](session var) varariable to control whether to rewrite OR to IN or not

### PR DESCRIPTION
**Pull request by Henry2SS**
…to in or not

# Proposed changes

Issue Number: close #12872   #15420

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java [hunk 1]
```diff
@@ -2238,5 +2238,14 @@ public void testRewriteOrToIn() throws Exception {
         sql = "SELECT * from test1 where (query_time = 1 or query_time = 2) and (scan_bytes = 2 or scan_bytes = 3)";
         explainString = UtFrameUtils.getSQLPlanOrErrorMsg(connectContext, "EXPLAIN " + sql);
         Assert.assertTrue(explainString.contains("PREDICATES: `query_time` IN (1, 2), `scan_bytes` IN (2, 3)"));
+
+        connectContext.getSessionVariable().setRewriteOrToInPredicateThreshold(100);
+        sql = "SELECT * from test1 where query_time = 1 or query_time = 2 or query_time in (3, 4)";
```
Comment (englefly) [inline fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java:2243]: REWRITE_OR_TO_IN_PREDICATE_THRESHOLD  包含 in的数量吗？

Comment (Henry2SS) [reply fe/fe-core/src/test/java/org/apache/doris/planner/QueryPlanTest.java:2243]: > REWRITE_OR_TO_IN_PREDICATE_THRESHOLD 包含 in的数量吗？

包含的。只要是 compound predicate 且 是OR连接的，都是包括了的，都算在exprs里面。

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.72 seconds
 load time: 700 seconds
 storage size: 17123035462 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221228113202_clickbench_pr_70378.html

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExtractCommonFactorsRule.java [hunk 2]
```diff
@@ -462,6 +464,13 @@ private Expr rewriteOrToIn(List<Expr> exprs) {
         boolean isOrToInAllowed = true;
         Set<String> slotSet = new LinkedHashSet<>();
 
+        int rewriteThreshold;
+        if (ConnectContext.get() == null) {
+            rewriteThreshold = 2;
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ExtractCommonFactorsRule.java:469]: is default value 2 too small?

Comment (Henry2SS) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/ExtractCommonFactorsRule.java:469]: Default value is set to 2, it means rewrite will always be allowed.

**Review by englefly — APPROVED** [review]

**Review by caiconghui — APPROVED** [review]
LGTM

**Pull request merged**